<!doctype html><html><head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Scientific Programming with Python</title>

<style type='text/css'>
  *{margin:0;padding:0;}
  body {
    font:13.34px helvetica,arial,freesans,clean,sans-serif;
    color:black;
    line-height:1.4em;
    background-color: #F8F8F8;
    padding: 0.7em;
  }
  p {
    margin:1em 0;
    line-height:1.5em;
  }
  table {
    font-size:inherit;
    font:100%;
    margin-top:1em;
    margin-bottom:1em;
  }
  table th{border-bottom:1px solid #bbb;padding:.2em 1em;}
  table td{border-bottom:1px solid #ddd;padding:.2em 1em;}
  input[type=text],input[type=password],input[type=image],textarea{font:99% helvetica,arial,freesans,sans-serif;}
  select,option{padding:0 .25em;}
  optgroup{margin-top:.5em;}
  pre,code{font:12px Monaco,"Courier New","DejaVu Sans Mono","Bitstream Vera Sans Mono",monospace;}
  pre {
    margin:1em 0;
    font-size:12px;
    background-color:#eee;
    border:1px solid #ddd;
    padding:5px;
    line-height:1.5em;
    color:#444;
    overflow:auto;
    -webkit-box-shadow:rgba(0,0,0,0.07) 0 1px 2px inset;
    -webkit-border-radius:3px;
    -moz-border-radius:3px;border-radius:3px;
  }
  pre code {
    padding:0;
    font-size:12px;
    background-color:#eee;
    border:none;
  }
  code {
    font-size:12px;
    background-color:#f8f8ff;
    color:#444;
    padding:0 .2em;
    border:1px solid #dedede;
  }
  img{border:0;max-width:100%;}
  abbr{border-bottom:none;}
  a{color:#4183c4;text-decoration:none;}
  a:hover{text-decoration:underline;}
  a code,a:link code,a:visited code{color:#4183c4;}
  h2,h3{margin:1em 0;}
  h1,h2,h3,h4,h5,h6{border:0;}
  h1{font-size:170%;border-top:4px solid #aaa;padding-top:.5em;margin-top:1.5em;}
  h1:first-child{margin-top:0;padding-top:.25em;border-top:none;}
  h2{font-size:150%;margin-top:1.5em;border-top:4px solid #e0e0e0;padding-top:.5em;}
  h3{margin-top:1em;}
  hr{border:1px solid #ddd;}
  ul{margin:1em 0 1em 2em;}
  ol{margin:1em 0 1em 2em;}
  ul li,ol li{margin-top:.5em;margin-bottom:.5em;}
  ul ul,ul ol,ol ol,ol ul{margin-top:0;margin-bottom:0;}
  blockquote{margin:1em 0;border-left:5px solid #ddd;padding-left:.6em;color:#555;}
  dt{font-weight:bold;margin-left:1em;}
  dd{margin-left:2em;margin-bottom:1em;}
  sup {
      font-size: 0.83em;
      vertical-align: super;
      line-height: 0;
  }
  * {
    -webkit-print-color-adjust: exact;
  }
  @media screen and (min-width: 914px) {
      body {
          width: 854px;
          margin:0 auto;
      }
  }
  @media print {
    table, pre {
      page-break-inside: avoid;
    }
    pre {
      word-wrap: break-word;
    }
  }

  .highlight {
    background-color:#f1c40f;
    padding:1em;
    
  }
</style>

</head>
<h1 id="cmps-130-homework-4">CMPS 130 - Homework #4</h1>
<p><img src="quad.gif" style="margin-left:1em;margin-bottom:1em;float:right"/><br>Our genome - DNA/RNA - are made up of long strings of nucleotides - guanine (G), adenine (A), thymine (T), or cytosine (C).  Genes are subsequences of our genome responsible for producting the specific protiens that make up our bodies.  Genes are typically between 2000 and 10000 nucleotides long.</p>
<p>Within genes, there exist even smaller subsequences - called <em>motifs</em>.  Motifs might be very short - <code>TTATT</code>, or longer and more complex, such as a <a href="http://en.wikipedia.org/wiki/G-quadruplex">G-quadruplex</a> - <code>GGATGAGCCTGTGGGTGGAGGGCAGTGGAGCGGGCTGGG</code>.  These motifs often perform very specific biological functions, such as controlling protein production.</p>
<p>The human genome has been mapped, and there are downloadable databases of text files containing long strings of G&#39;s, A&#39;s, T&#39;s, and C&#39;s representing our genes (and those of many other organisms).  In Bioinformatics, computer programs are used to analyze and search this genetic data for patterns and motifs.</p>
<p>Since most of our genetic material is <a href="http://en.wikipedia.org/wiki/Noncoding_DNA">non-functional</a> (or at least, <a href="http://healthland.time.com/2012/09/06/junk-dna-not-so-useless-after-all/">so we thought</a>), an important aspect of using computational tools for classifying motifs is determining whether motifs are occuring at random, or because they are serving a real purpose.  A key method bioinformatics researchers use is to calculate the <em>likelhood</em> of finding a specific motif among a <strong>random</strong> sequence.  If a specific motif occurs in the genome far more frequently than the <em>expected</em> frequency, we begin to assume the motif is doing something important.  It&#39;s at that point that real &quot;wet lab&quot; tests will begin to actually investigate the specific motif in more detail.</p>
<p>For this assignment you will write a program that would help a scientists determine if a specific motif is worth investigating by reporting its <em>expected</em> frequency in a number of different ways.</p>
<h2 id="part-1-30-points-">Part 1 (30 points)</h2>
<p>Your program should start by asking the user to enter a specific motif - a sting of A, T, C, G&#39;s.  Your program should ensure that only ATCG combinations are permitted, if the user enters an invalid string allow them to try again until they do.</p>
<p>It turns out that A, T, C, and G do not appear at the same frequency in the genome, and their distributions vary widely across different genes.  Next, ask the user to enter the frequency at which each nucleotide exists in their hypothetical genome.  </p>
<p>Using those frequencies, calculate the likely hood that a <em>single</em> sequence of the <em>same</em> length is precisely the motif that the user entered.</p>
<p>For example - if the frequencies are as follows:  A= 20%, T = 28%, C = 23% and G = 29%, the following motif has a likelyhood of appearing:</p>
<pre><code>Motif:  CGGA
.23 * .29 * .29 * .20 = 0.0038686
</code></pre><p>Meaning if you were to pick 4 nucleotides at complete random, you&#39;d expect to see CGGA roughly 0.4% of the time. This makes intuitive sense, since there are 4 nucleotides and 4 options (A T C or G), there are 4 <sup>4</sup> = 256 combinations.  1 / 256 is about 0.0039, and we see a slightly different likelyhood for our motif because the frequency of A, T, C, and G are <strong>not</strong> equal.</p>
<p>With a different set of frequencies we&#39;d get a very different number:</p>
<pre><code>A = 30%
T = 5%
G = 40%
C = 25%

Motif: CGGA
.25 * .40 * .40 * .30 = 0.012 (or 1.2%)
</code></pre><h2 id="part-2-30-points-">Part 2 (30 points)</h2>
<p>The likelhood of a motif appearing at one specific location seems really low - because it is!  However, because genes typically range between 2000 and 10000 nucleotides, the chance that you find a given motif at least once <strong>anywhere</strong> in the gene is actually a bit higher.  If you have a gene of 5000 nucleotides, you essentially have 5000 opportunities to start one of these motifs!  So the likelyhood of finding the motif from part one (using the first set of frequencies) is actually <code>0.0038686 * 5000</code> - or 7.1%</p>
<p>Using PyLab (matplotlib) create a plot that shows the user probability of finding at least one of their motifs in sequences ranging in length from N to 10,000 - where N is the length of the motif they entered.  The x-axis should be the sequence length and the y-axis should be the probability of finding the sequence at least once.</p>
<h2 id="part-3-40-points-total-">Part 3 (40 points - total)</h2>
<p>Another important measure a bioinformatics researcher will employ is <strong>enrichment</strong>.  Enrichment is the <strong>number</strong> of times a motif is found within a sequence.  While our calculation in Part 2 tells us the likely hood of a motif occuring <em>at least once</em>, in this part we&#39;ll use a Monte Carlo simulation to tell us exactly how many times we should expect to find the motif in a sequence.</p>
<h3 id="part-3a-20-points-">Part 3a (20 points)</h3>
<p>For this part, you&#39;ll first need to write some code (probably a function) that generates a random sequence of length N based on the frequency distributions for A, T, C, and G the user entered.  Here&#39;s how:</p>
<p>Lets say the frequencies are as follows:  A= 20%, T = 28%, C = 23% and G = 29%</p>
<p>For each nucleotide in your sequence, create a random floating point number (<code>import random ... random.random()</code>).  If the value is &lt; .20, then assign A. If the value is &gt;= .2 and &lt; 0.48, then assign T.  If the value is &gt;= 0.48 and &lt; 0.71, assign C.  Finally, if the value is &gt; =.71, assign G.  Essentially, you are dividing the number line of real numbers between 0 and 1.0 into regions proportional to the frequencies, and then assigning the letter by which region the random number fall into.</p>
<h3 id="part-3b-20-points-">Part 3b (20 points)</h3>
<p>Now generate a plot for N = 100, N = 1000, N = 2000, N = 5000, and N = 10000 (where N is the length of the sequence).  For each of these values, generate 100 random sequences of the given length and count how many times you find the motif the user entered in each random sequence you generate.  The y-axis of the plot will be the <em>average</em> number of times you found the motif, for the given sequence length, in the 100 random samples you generated.</p>
<p>Note - its easy to count the number of motifs that occur in a sequence - you can use the string&#39;s <code>find</code> method.  You can read about it <a href="https://docs.python.org/release/3.1.3/library/stdtypes.html#str.find">here</a>.</p>
<p>The <code>find</code> method accepts a string to search for within a string.  It returns the (first) index where that substring was found, and returns -1 if it was not found.  <code>find</code> also accepts a starting position.  So, to count how many times a substring <code>sub</code> is found in a string <code>str</code> you&#39;d call <code>str.find(sub, start)</code>.  The first time you call this start will be zero.  If the method returns -1, <code>sub</code> doesn&#39;t appear in <code>str</code>.  If the method returns some index <code>i</code> then <code>sub</code> is in <code>str</code> once.  You then set <code>start</code> to <code>i+1</code> and recall <code>str.find(sub, start)</code>.  Repeat this process until the find method does indeed return -1 - and keep count of how many times it returns a value &gt;= 0.</p>
<h3 id="grading">Grading</h3>
<p>I&#39;ll run your code with a few test cases, but here is one that will help you understand if you have the right values.</p>
<p><pre><br>Motif:   AGGAGGTGGGG</p>
<table>
<thead>
<tr>
<th>Nucleotide</th>
<th>Frequency</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>20%</td>
</tr>
<tr>
<td>T</td>
<td>27%</td>
</tr>
<tr>
<td>C</td>
<td>19%</td>
</tr>
<tr>
<td>G</td>
<td>34%</td>
</tr>
</tbody>
</table>
<p>Probability of appearing in a sequence N = 11 is 0.00000192865742 (0.000192865742%)</p>
<p>Plot of probability of finding at least one of these motifs in a sequences of length N to 10,000</p>
<p>Expected number of motifs in sequences of length N = 100, N = 1000, N = 2000, N = 5000, and N = 10000<br></pre><br>```</p>
</html>