<!doctype html><html><head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Scientific Programming with Python</title>

<style type='text/css'>
  *{margin:0;padding:0;}
  body {
    font:13.34px helvetica,arial,freesans,clean,sans-serif;
    color:black;
    line-height:1.4em;
    background-color: #F8F8F8;
    padding: 0.7em;
  }
  p {
    margin:1em 0;
    line-height:1.5em;
  }
  table {
    font-size:inherit;
    font:100%;
    margin-top:1em;
    margin-bottom:1em;
  }
  table th{border-bottom:1px solid #bbb;padding:.2em 1em;}
  table td{border-bottom:1px solid #ddd;padding:.2em 1em;}
  input[type=text],input[type=password],input[type=image],textarea{font:99% helvetica,arial,freesans,sans-serif;}
  select,option{padding:0 .25em;}
  optgroup{margin-top:.5em;}
  pre,code{font:12px Monaco,"Courier New","DejaVu Sans Mono","Bitstream Vera Sans Mono",monospace;}
  pre {
    margin:1em 0;
    font-size:12px;
    background-color:#eee;
    border:1px solid #ddd;
    padding:5px;
    line-height:1.5em;
    color:#444;
    overflow:auto;
    -webkit-box-shadow:rgba(0,0,0,0.07) 0 1px 2px inset;
    -webkit-border-radius:3px;
    -moz-border-radius:3px;border-radius:3px;
  }
  pre code {
    padding:0;
    font-size:12px;
    background-color:#eee;
    border:none;
  }
  code {
    font-size:12px;
    background-color:#f8f8ff;
    color:#444;
    padding:0 .2em;
    border:1px solid #dedede;
  }
  img{border:0;max-width:100%;}
  abbr{border-bottom:none;}
  a{color:#4183c4;text-decoration:none;}
  a:hover{text-decoration:underline;}
  a code,a:link code,a:visited code{color:#4183c4;}
  h2,h3{margin:1em 0;}
  h1,h2,h3,h4,h5,h6{border:0;}
  h1{font-size:170%;border-top:4px solid #aaa;padding-top:.5em;margin-top:1.5em;}
  h1:first-child{margin-top:0;padding-top:.25em;border-top:none;}
  h2{font-size:150%;margin-top:1.5em;border-top:4px solid #e0e0e0;padding-top:.5em;}
  h3{margin-top:1em;}
  hr{border:1px solid #ddd;}
  ul{margin:1em 0 1em 2em;}
  ol{margin:1em 0 1em 2em;}
  ul li,ol li{margin-top:.5em;margin-bottom:.5em;}
  ul ul,ul ol,ol ol,ol ul{margin-top:0;margin-bottom:0;}
  blockquote{margin:1em 0;border-left:5px solid #ddd;padding-left:.6em;color:#555;}
  dt{font-weight:bold;margin-left:1em;}
  dd{margin-left:2em;margin-bottom:1em;}
  sup {
      font-size: 0.83em;
      vertical-align: super;
      line-height: 0;
  }
  * {
    -webkit-print-color-adjust: exact;
  }
  @media screen and (min-width: 914px) {
      body {
          width: 854px;
          margin:0 auto;
      }
  }
  @media print {
    table, pre {
      page-break-inside: avoid;
    }
    pre {
      word-wrap: break-word;
    }
  }

  .highlight {
    background-color:#f1c40f;
    padding:1em;
    
  }
</style>

</head>
<h1 id="cmps-130-homework-cs-elective-requirement">CMPS 130 - Homework CS Elective Requirement</h1>
<p><img src="http://newsoffice.mit.edu/sites/mit.edu.newsoffice/files/styles/news_article_image_top_slideshow/public/images/2015/MIT-Quad-Genome.jpg?itok=qMwDyhOy" style="float:left; margin-right:1em; width:300px"/><br>If you are a Computer Science major wishing to count this course as one of your seven electives, you are required to complete the following additional assignment before the last day of the class.  As a computer science major, you already know C and C++ well (perhaps C++ more than C).  In this assignment you will implement a well known algorithm in computer science commonly used in genetics research in both Python and C++.  You will also evaluate the two, for both runtime efficiency and ease of programming.  The purpose of this assignment is for you to better understand the similarities, differences, and relative strengths of Python and C++ on a common problem.</p>
<h2 id="the-algorithm">The Algorithm</h2>
<p>You will implement an algorithm that calculates the <em>edit distance</em>, or more specifically, the <a href="https://en.wikipedia.org/wiki/Levenshtein_distance">Levenshtein_distance</a> between two strings.  Tthe Levenshtein distance between two strings is the minimum number of single-character edits (i.e. insertions, deletions or substitutions) required to change one string into the other.  This is particularly useful when studying the similarity between genes from different organisms.  DNA sequences conserved (small edit distance) across species are likely to be functionally important, while variations between members of the same species can indicate different susceptibilities to disease.</p>
<p>For example, we might examine the following short DNA sequences:</p>
<pre><code>GGAAGGGGCGATCGGAGGGC
GGTAAGGGGCCTGATCGAAGGGCAA
</code></pre><p>We can calculate the edit distance, and record insertions/deletions/substitutions in such a way that we can print a <em>aligned</em> set of sequences, with dashes representing misaligned symbols</p>
<pre><code>GG-AAGGGG-C-GATCGGAGGGC--
GGTAAGGGGCCTGATCGAAGGGCAA
</code></pre><p>There are several ways of calculating edit distance.  The widely accepted <em>best</em> algorithm is the Wagnerâ€“Fischer algorithm, which utilizes a 2-dimensional array to calculate the answer in O(N&dot;M) time, and O(N&dot;M) memory space - where N and M are the lengths of the two strings being compared.  When the strings are of similar length, you can think of the algorithm being O(N<sup>2</sup>).  No one has ever said O(N<sup>2</sup>) is fast (or efficient), but since all the other known algorithms are <em>exponential</em>, the Wagner-Fischer is widely adopted.  Computer scientists have been debating for decades whether there exists a better (more efficient) algorithm for this type of problem - but it has <a href="http://newsoffice.mit.edu/2015/algorithm-genome-best-possible-0610">recently been proven to be optimal</a>!</p>
<h2 id="calculating-edit-distance">Calculating Edit Distance</h2>
<pre><code>* Input:  Two strings, S and T
* Construct an M x N matrix (array) of integers in memory, called D, 
  where M is equal to the length of S + 1, and N is equal to the length of T + 1. 
* Assign the values 0...M to the first row of the array, and 0..N to the first column.
</code></pre><h3 id="processing">Processing</h3>
<p>Starting at element [1,1], proceed down each column, working your way to the right <em>after</em> completing each column.  </p>
<pre><code>At each cell [i,j]:
* if S[j] is equal to T[i], then assign D[i,j] to be the value D[i-1, j-1].
* Otherwise, assign D[i, j] to be the **minimum** of the following three values:
    * D[i-1, j] + 1  (representing a deletion)
    * D[i, j-1] + 1  (representing an insertion)
    * D[i-1, j-1] + 1 (representing a substitution)
</code></pre><h3 id="result">Result</h3>
<p>Once you&#39;ve processed all the cells, the bottom right-most cell contains the (optimal) total number of edits to transform S into T.</p>
<h2 id="recording-the-actual-edits">Recording the actual edits</h2>
<p>The algorithm above is fairly straightforward, however it only gives us the end result - the edit distance.  When comparing genetic sequences, its usually important to use the edit operations to construct a <em>sequence alignment</em>.  To do this, we not only need to create the matrix described above, but we need to record <em>which</em> operations would be performed along the way.</p>
<p>To do this, you should make an auxilary array, E, of the same dimensions as D in the description above.  When assigning D[i, j], assign E[i, j] as the following:</p>
<pre><code>* if S[j] is equal to T[i], then assign E[i, j] to be 0, 
  (representing the fact that the value is derived from the cell diagnoal from this cell (above and left).
* Otherwise, assign E[i, j] according to which D cell is used to derive the value of D[i, j] as follows:
    * If D[i-1, j] was used, record 1 (derived from cell above)
    * If D[i, j-1] was used, record 2 (derived from cell to the left)
    * If D[i-1, j-1] was used, record 0 (derived from cell diagonal to this one)
</code></pre><h2 id="backtracing-to-obtain-sequences-with-gaps-inserted">Backtracing to obtain sequences with gaps inserted</h2>
<p>Once D and E have been constructed, you can now create <em>aligned</em> sequences, representing insert/deletes with gaps (-) in the corresponding sequences by <em>backtracing</em> E. </p>
<p>Start with two empty strings, representing the final S and T string.  Lets call them S_RESULT and T_RESULT</p>
<p>Starting at the lower right of E (the same cell you got the edit distance from D), proceed as follows:</p>
<pre><code>* if the cell E[i, j] is 0, **prepend** S[j-1] to S_RESULT, **prepend** T[i-1] to T_RESULT, and decrement i and j
* if the cell E[i, j] is 1, **prepend** S_RESULT with a &#39;-&#39;, **prepend** T[i-1] to T_RESULT, and decrement i
* if the cell E[i, j] is 2, **prepend** S[j-1] to S_RESULT, **prepend** T_RESULT with a &#39;-&#39;, and decrement j.
</code></pre><p>Depending if you are using C++ or Python, this string building exercise might be done slightly different - but the point is that when you see a cell derived from the diagonal, you use characters from both inputs.  On a cell derived from the left, you place a gap in the up/down string, and if you see a cell derived from above, you place a gap in the left/right string.  You are building the string in reverse, and when r or c gets to 0, use the remaining characters from either the left/right or up/down string to finish off the process.</p>
<h1 id="your-assignment">Your Assignment</h1>
<p>For this assignment, you will write two programs that do the exact same thing - one in C++ and one in Python.</p>
<p>Your programs should first accept two strings on the command line (see below for help with command line arguments), which represent the filename of two files.  Each of these files should just contain simple strings, I&#39;ve provided three sample sets - the first two are very small test files, helpful during debugging.  The third pair represents the NRAS genomic sequence in humans and in cows (these genes are involved in some cancers).</p>
<ul>
<li>Sample set 1 - <a href="s1.txt">S</a>, <a href="t1.txt">T</a></li>
<li>Sample set 2 - <a href="s2.txt">S</a>, <a href="t2.txt">T</a></li>
<li>Sample set 3 - <a href="nras_human.txt">Human NRAS</a>, <a href="nras_cow.txt">Cow NRAS</a></li>
</ul>
<p>Your program should open the files based on the command line arguments and read the strings (remove new line characters) into variables.</p>
<p>Next, perform the Wagner-Fischer computation, including the backtracing procedure explained above.  </p>
<p>Your program should print out the resulting aligned sequences/strings, and also print out the number of edits (the edit distance). You should also print the total elapsed run time of the processing.  See below for details on timing in C++, you already know how to do this in Python.</p>
<p>Here is output of the first two samples.  Your program should have the exact same results, except of course for the elapsed time, which will be unique for your machine.  I compiled my C++ solution into a program named <code>editc</code>.  The Python version would print the same results (with a different time)</p>
<pre><code>&gt; editc s1.txt s2.txt
Input Sequences
----------------------------
Saturday
Sunday
----------------------------
Aligned Sequences
----------------------------
Saturday
S--unday
----------------------------
The minimum edit distance is 3
Completed in 0.000100 seconds.
</code></pre><pre><code>&gt; editc s2.txt s2.txt
Input Sequences
----------------------------
GGAAGGGGCGATCGGAGGGC
GGTAAGGGGCCTGATCGAAGGGCAA
----------------------------
Aligned Sequences
----------------------------
GG-AAGGGG-C-GATCGGAGGGC--
GGTAAGGGGCCTGATCGAAGGGCAA
----------------------------
The minimum edit distance is 6
Completed in 0.000129 seconds.
</code></pre><h2 id="testing-and-debugging">Testing and Debugging</h2>
<p>Unless you have unusually impressive skills, you might not get this to work the first time your run your program.   Below is the actual matrix that would be generated for Samples 1 and 2.  You might want to print yours out, so you can compare to see what could be going wrong..</p>
<p>Sample 1</p>
<pre><code>       S  a  t  u  r  d  a  y
=============================
  | 0  1  2  3  4  5  6  7  8  
S | 1  0  1  2  3  4  5  6  7  
u | 2  1  1  2  2  3  4  5  6  
n | 3  2  2  2  3  3  4  5  6  
d | 4  3  3  3  3  4  3  4  5  
a | 5  4  3  4  4  4  4  3  4  
y | 6  5  4  4  5  5  5  4  3
</code></pre><p>Sample 2</p>
<pre><code>
         G   G   A   A   G   G   G   G  C    G   A   T   C   G   G   A   G   G   G   C
======================================================================================
  | 0    1   2   3   4   5   6   7   8  9   10  11  12  13  14  15  16  17  18  19  20
G | 1    0   1   2   3   4   5   6   7  8    9  10  11  12  13  14  15  16  17  18  19    
G | 2    1   0   1   2   3   4   5   6  7    8   9  10  11  12  13  14  15  16  17  18
T | 3    2   1   1   2   3   4   5   6  7    8   9   9  10  11  12  13  14  15  16  17    
A | 4    3   2   1   1   2   3   4   5  6    7   8   9  10  11  12  12  13  14  15  16  
A | 5    4   3   2   1   2   3   4   5  6    7   7   8   9  10  11  12  13  14  15  16  
G | 6    5   4   3   2   1   2   3   4  5    6   7   8   9   9  10  11  12  13  14  15  
G | 7    6   5   4   3   2   1   2   3  4    5   6   7   8   9   9  10  11  12  13  14  
G | 8    7   6   5   4   3   2   1   2  3    4   5   6   7   8   9  10  10  11  12  13  
G | 9    8   7   6   5   4   3   2   1  2    3   4   5   6   7   8   9  10  10  11  12  
C | 10   9   8   7   6   5   4   3   2  1    2   3   4   5   6   7   8   9  10  11  11  
C | 11  10   9   8   7   6   5   4   3  2    2   3   4   4   5   6   7   8   9  10  11  
T | 12  11  10   9   8   7   6   5   4  3    3   3   3   4   5   6   7   8   9  10  11  
G | 13  12  11  10   9   8   7   6   5  4    3   4   4   4   4   5   6   7   8   9  10  
A | 14  13  12  11  10   9   8   7   6  5    4   3   4   5   5   5   5   6   7   8   9  
T | 15  14  13  12  11  10   9   8   7  6    5   4   3   4   5   6   6   6   7   8   9  
C | 16  15  14  13  12  11  10   9   8  7    6   5   4   3   4   5   6   7   7   8   8  
G | 17  16  15  14  13  12  11  10   9  8    7   6   5   4   3   4   5   6   7   7   8  
A | 18  17  16  15  14  13  12  11  10  9    8   7   6   5   4   4   4   5   6   7   8  
A | 19  18  17  16  15  14  13  12  11  10   9   8   7   6   5   5   4   5   6   7   8  
G | 20  19  18  17  16  15  14  13  12  11  10   9   8   7   6   5   5   4   5   6   7  
G | 21  20  19  18  17  16  15  14  13  12  11  10   9   8   7   6   6   5   4   5   6  
G | 22  21  20  19  18  17  16  15  14  13  12  11  10   9   8   7   7   6   5   4   5  
C | 23  22  21  20  19  18  17  16  15  14  13  12  11  10   9   8   8   7   6   5   4  
A | 24  23  22  21  20  19  18  17  16  15  14  13  12  11  10   9   8   8   7   6   5  
A | 25  24  23  22  21  20  19  18  17  16  15  14  13  12  11  10   9   9   8   7   6
</code></pre><h2 id="implementation-in-c-">Implementation in C++</h2>
<p>If you haven&#39;t used command line arguments before in C++, here&#39;s a code snippet to help you:</p>
<pre><code>#include &lt;iostream&gt;
using namespace std;

int main(int argv, char ** argc) {
    if ( argv &gt;= 3 ) {
        cout &lt;&lt; argc[1] &lt;&lt; endl;
        cout &lt;&lt; argc[2] &lt;&lt; endl;
    }
}
</code></pre><p>If you run your program from the command line (lets say you compile it to be named <code>test</code>), you can enter two arguments and they will print to the screen.  These would be the filenames in your actual program...</p>
<pre><code>&gt; test hello world
hello
world
&gt;
</code></pre><p>You may use and valid C or C++ to open the text files and read them into strings.  You can use any method of keeping track of time that you find convenient - here is a sample of how to do this with the standard C library.</p>
<pre><code>#include &lt;ctime&gt;
#include &lt;iostream&gt;
using namespace std;

int main() {
    int start = clock();

    for ( int i = 0; i &lt; 10000; i++ ) {
        cout &lt;&lt; &quot;.&quot;;
    }
    cout &lt;&lt; endl;

    int t = clock() - start;
    cout &lt;&lt; &quot;Completed in &quot; &lt;&lt;((float)t)/CLOCKS_PER_SEC &lt;&lt; &quot; seconds.&quot; &lt;&lt; endl;;
}
</code></pre><h2 id="implementation-in-python">Implementation in Python</h2>
<p>We haven&#39;t used command line arguments in python yet - but its really easy:</p>
<pre><code>import sys

for i in sys.argv[1:]:
    print(i)
</code></pre><p>Again, if you run your program (test.py), and enter command line arguments, they&#39;ll print to the screen</p>
<pre><code>&gt; python3 test.py hello world
hello
world
&gt;
</code></pre><p>I will assume you know all about two-dimensional arrays in C++, but we haven&#39;t discussed them in Python.  There is no real concept for N-dimensional arrays built into Python (although <a href="http://www.numpy.org/">numpy</a> adds them).  Instead, you must use a <em>list of lists</em>.</p>
<p>For example, you can create a 5x5 matrix in Python using the following syntax (all initialized to 0&#39;s)</p>
<pre><code>rows = 5
cols = 5
matrix = [[0 for x in range(cols)] for x in range(rows)]
</code></pre><p>You can then access each element using standard bracket notation:</p>
<pre><code>matrix[3][2] = 1
print(matrix)
</code></pre><p>The code above prints <code>[[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 1, 0, 0], [0, 0, 0, 0, 0]]</code></p>
<h2 id="grading">Grading</h2>
<table>
<thead>
<tr>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>C++ Edit Distance</td>
<td>30%</td>
</tr>
<tr>
<td>Python Edit Distance</td>
<td>30%</td>
</tr>
<tr>
<td>C++ Backtrace/Alignment Printout</td>
<td>15%</td>
</tr>
<tr>
<td>Python Backtrace/Alignment Printout</td>
<td>15%</td>
</tr>
<tr>
<td>Overall Code Quality</td>
<td>10%</td>
</tr>
</tbody>
</table>
</html>